GVFS/GVFS.UnitTests/Common/LibGit2RepoInvokerTests.cs

```diff
@@ -25,32 +23,65 @@ public void Setup()
             this.invoker?.Dispose();
 
             this.tracer = new MockTracer();
-            this.invoker = new LibGit2RepoInvoker(this.tracer, this.CreateRepo, this.disposalPeriod);
             this.numConstructors = 0;
             this.numDisposals = 0;
             this.DisposalTriggers = new BlockingCollection<object>();
+
+            this.invoker = new LibGit2RepoInvoker(this.tracer, this.CreateRepo);
         }
 
         [TestCase]
-        public void DoesNotCreateRepoOnConstruction()
+        public void DoesCreateRepoOnConstruction()
         {
-            this.numConstructors.ShouldEqual(0);
+            this.numConstructors.ShouldEqual(1);
         }
 
         [TestCase]
-        public void CreatesRepoOnTryInvoke()
+        public void CreatesOnRequestAfterClosed()
         {
-            this.numConstructors.ShouldEqual(0);
+            this.numConstructors.ShouldEqual(1);
+
+            this.invoker.DisposeSharedRepo();
+
+            this.numDisposals.ShouldEqual(1);
+            this.numConstructors.ShouldEqual(1);
+
+            this.invoker.InitializedSharedRepo();
+
+            this.numDisposals.ShouldEqual(1);
+            this.numConstructors.ShouldEqual(2);
 
             this.invoker.TryInvoke(repo => { return true; }, out bool result);
-            result.ShouldEqual(true);
+
+            this.numDisposals.ShouldEqual(1);
+            this.numConstructors.ShouldEqual(2);
+        }
+
+        [TestCase]
+        public void CreatesOnInvokeAfterClosed()
+        {
             this.numConstructors.ShouldEqual(1);
+
+            this.invoker.DisposeSharedRepo();
+
+            this.numDisposals.ShouldEqual(1);
+            this.numConstructors.ShouldEqual(1);
+
+            this.invoker.TryInvoke(repo => { return true; }, out bool result);
+
+            this.numDisposals.ShouldEqual(1);
+            this.numConstructors.ShouldEqual(2);
+
+            this.invoker.InitializedSharedRepo();

```

Comment:
Maybe add a comment here that `InitializedSharedRepo` should not construct another libgit2 if after one was created for TryInvoke?